eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumAddressStrategy.scala

```diff
@@ -0,0 +1,172 @@
+    tx.copy(txIn = utxos.map { case utxo =>
+      // we use dummy signature here, because the result is only used to estimate fees
+      val sig = ByteVector.fill(71)(1)
+      val sigScript = Script.write(OP_PUSHDATA(Script.write(Script.pay2wpkh(utxo.key.publicKey))) :: Nil)
+      val witness = ScriptWitness(sig :: utxo.key.publicKey.value :: Nil)
+      TxIn(utxo.outPoint, signatureScript = sigScript, sequence = TxIn.SEQUENCE_FINAL, witness = witness)
+    })
+  }
+
+  override def extractPubKey(txIn: TxIn): Option[PublicKey] = {
+    Try {
+      // we're looking for tx that spend a pay2sh-of-p2wkph output
+      require(txIn.witness.stack.size == 2)
+      val sig = txIn.witness.stack(0)
+      val pub = txIn.witness.stack(1)
+      val OP_PUSHDATA(script, _) :: Nil = Script.parse(txIn.signatureScript)
+      val publicKey = PublicKey(pub)
+      if (Script.write(Script.pay2wpkh(publicKey)) == script) {
+        Some(publicKey)
+      } else None
+    } getOrElse None
+  }
+
+  override def computeRootPub(master: ExtendedPrivateKey, chainHash: ByteVector32): (String, String) = {
+    val xpub = DeterministicWallet.publicKey(DeterministicWallet.derivePrivateKey(master, bip49RootPath(chainHash)))
+    val prefix = chainHash match {
+      case Block.LivenetGenesisBlock.hash => DeterministicWallet.ypub
+      case Block.RegtestGenesisBlock.hash | Block.TestnetGenesisBlock.hash => DeterministicWallet.upub
+    }
+    (DeterministicWallet.encode(xpub, prefix), xpub.path.toString())
+  }
+}
+
+class NativeSegwitStrategy extends ElectrumAddressStrategy {
+
+  /**
+    * @param key the public key
+    * @return the bech32 encoded witness program for the p2wpkh script of this key
+    */
+  override def computeAddress(key: PublicKey, chainHash: ByteVector32): String = chainHash match {
+    case Block.RegtestGenesisBlock.hash => Bech32.encodeWitnessAddress("bcrt", 0, Crypto.hash160(key.value))
+    case Block.TestnetGenesisBlock.hash => Bech32.encodeWitnessAddress("tb", 0, Crypto.hash160(key.value))
+    case Block.LivenetGenesisBlock.hash => Bech32.encodeWitnessAddress("bc", 0, Crypto.hash160(key.value))
+  }
+
+  override def computePublicKeyScript(key: PublicKey) = Script.pay2wpkh(key)
+
+  override def signTx(tx: Transaction, d: Data): Transaction = {
+    tx.copy(txIn = tx.txIn.zipWithIndex.map { case (txIn, i) =>
+      val utxo = d.utxos.find(_.outPoint == txIn.outPoint).getOrElse(throw new RuntimeException(s"cannot sign input that spends from ${txIn.outPoint}"))
+      val key = utxo.key
+      val sig = Transaction.signInput(tx, i, Script.pay2pkh(key.publicKey), SIGHASH_ALL, Satoshi(utxo.item.value), SigVersion.SIGVERSION_WITNESS_V0, key.privateKey)
+      val sigScript = ByteVector.empty
+      val witness = ScriptWitness(sig :: key.publicKey.value :: Nil)
+      txIn.copy(signatureScript = sigScript, witness = witness)
+    })
+  }
+
+  override def addUtxos(tx: Transaction, utxos: Seq[Utxo]): Transaction = {
```

Comment:
We should keep the previous name `addUtxosWithDummySig` which is more explicit (here is not obvious that the output tx contains dummy wittness data)